### ISSUE TITLE
[chain] Revised Chain Storage

### ISSUE THREAD

**Issue by zivkovicmilos** [gnolang/gno#1825]
## Description

This effort covers the concern of storage creep in Gnoland, as the blockchain advances to new heights.

Essentially, as the Gno chain progresses, it will become more and more expensive to manage storage-wise, since it can grow into gigabytes / terabytes in size for archive nodes. This can be mitigated with a few strategies:
- introduce different node types, so not all nodes are archive nodes (this is an entirely new effort, and not something required for main-net)
- analyze and optimize the data structures used to store blockchain data in embedded storage
- analyze and optimize the underlying storage layer (embedded storage)

Successful outcome of this effort:
- storage growth contained in the form of optimizations
- storage growth analyzed and documented
- storage data structure and underlying layer analyzed and optimized to minimize overhead of a growing chain

**Comment by moul** [comment]
This is definitely a significant challenge for us, but we will primarily address it after the launch by introducing special node types, among other things. 

Before the launch, we should consider implementing "storage rent gas" for contracts. However, the question is how to establish proper garbage collection, which depends on #2743. 

Let's remove "must have" for now.

**Comment by moul** [comment]
More info about gas storage fee mechanism here: #3418
